spring-kafka/src/main/java/org/springframework/kafka/support/serializer/DelegatingByTopicDeserializer.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2019-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.support.serializer;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
+
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.serialization.Deserializer;
+
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * A {@link Deserializer} that delegates to other deserializers based on the topic name.
+ *
+ * @author Gary Russell
+ * @since 2.8
+ *
+ */
+public class DelegatingByTopicDeserializer implements Deserializer<Object> {
+
+	private final Map<Pattern, Deserializer<? extends Object>> delegates = new ConcurrentHashMap<>();
+
+	private boolean forKeys;
+
+	private boolean cased = true;
+
+	/**
+	 * Construct an instance that will be configured in {@link #configure(Map, boolean)}
+	 * with consumer properties.
+	 */
+	public DelegatingByTopicDeserializer() {
+	}
+
+	/**
+	 * Construct an instance with the supplied mapping of topic name patterns to delegate
+	 * deserializers.
+	 * @param delegates the map of delegates.
+	 */
+	public DelegatingByTopicDeserializer(Map<Pattern, Deserializer<?>> delegates) {
+		this.delegates.putAll(delegates);
+	}
+
+	/**
+	 * Set to false to make topic name matching case insensitive.
+	 * @param caseSensitive false for case insensitive.
+	 */
+	public void setCaseSensitive(boolean caseSensitive) {
+		this.cased = caseSensitive;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void configure(Map<String, ?> configs, boolean isKey) {
+		this.forKeys = isKey;
+		Object insensitive = configs.get(DelegatingByTopicSerializer.CASE_SENSITIVE);
+		if (insensitive instanceof String) {
```

Comment:
Can it be a plain `boolean`, too?
Shouldn't we fail if it is some other unexpected type?